File: usage-daemon/src/providers/hyper.js
// Charm Hyper credits provider plugin.
//
// Two sources:
//   1. GET /v1/credits (Bearer API key) — returns {"balance": 78}
//   2. Billing page HTML (cookie) — parses "Next Hypercredit Refresh in 4 weeks"
//      to compute resets_at. Optional: if no cookie, resets_at is null.
//
// Free tier: 100 credits/month (hardcoded — API doesn't return cap).
// Used = cap - balance.

export const CREDITS_COLOR = '#0072B2'; // Okabe-Ito blue

export const ID = 'hyper';
export const LABEL = 'Charm Hyper';

const API_URL = 'https://hyper.charm.land/v1/credits';
const BILLING_URL = 'https://hyper.charm.land/teams/6b4e808f-3933-4bfd-a321-be1bcfe04f21/billing';
const USER_AGENT = 'usage-daemon/0.1';
const BROWSER_UA = 'Mozilla/5.0 (X11; Linux x86_64; rv:153.0) Gecko/20100101 Firefox/153.0';

const PLAN_CAP = 100; // free tier: 100 credits/month

export class AuthExpiredError extends Error {
  constructor(msg = 'Hyper API key missing or invalid') {
    super(msg);
    this.code = 'auth_expired';
  }
}

export class RateLimitedError extends Error {
  constructor(retryAfter = null) {
    super('rate_limited');
    this.code = 'rate_limited';
    this.retryAfter = retryAfter;
  }
}

// Parse "Next Hypercredit Refresh in 4 weeks" / "in 1 day" / "in 2 days"
// into an ISO timestamp. Returns null on failure.
export function parseRefreshText(html) {
  const m = html.match(/Next Hypercredit Refresh in (\d+)\s+(day|days|week|weeks)/i);
  if (!m) return null;
  const n = Number(m[1]);
  const unit = m[2].toLowerCase();
  let ms;
  if (unit.startsWith('week')) {
    ms = n * 7 * 24 * 3600 * 1000;
  } else {
    ms = n * 24 * 3600 * 1000;
  }
  return new Date(Date.now() + ms).toISOString();
}

// Pure function of the raw API JSON text — no fs/network.
export function parse(raw) {
  let data;
  try {
    data = typeof raw === 'string' ? JSON.parse(raw) : raw;
  } catch {
    throw new AuthExpiredError('unparseable hyper response');
  }
  if (!data || typeof data !== 'object') {
    throw new AuthExpiredError('unparseable hyper response');
  }

  // Auth failures surface as { error: { message, type } }.
  if (data.error) {
    if (data.error.type === 'authentication_error') {
      throw new AuthExpiredError(data.error.message || 'invalid hyper API key');
    }
    throw new Error(`hyper API error: ${data.error.message || JSON.stringify(data.error)}`);
  }

  const balance = data.balance;

  if (typeof balance !== 'number' || !Number.isFinite(balance)) {
    throw new AuthExpiredError('no usable hypercredits balance');
  }

  const used = Math.max(0, PLAN_CAP - balance);
  const pct = Math.max(0, Math.min(100, (100 * used) / PLAN_CAP));

  const windows = [
    {
      id: 'hypercredits',
      label: 'Credits',
      letter: 'Hc',
      pct,
      used,
      cap: PLAN_CAP,
      unit: 'credits',
      resets_at: null, // filled in by fetch() if billing page is reachable
      color: CREDITS_COLOR,
      will_deplete: false,
    },
  ];

  return {
    tier: 'free',
    windows,
    segments: [],
    // Raw balance for meta() — lets the UI show "78 remaining".
    _hyper: { balance },
  };
}

function createProvider() {
  let apiKey = null;
  let cookie = null;

  return {
    id: ID,
    label: LABEL,
    auth: { kind: 'token' },

    config() {
      return {
        id: ID,
        label: LABEL,
        usageUrl: 'https://hyper.charm.land/docs/api/credits.html#pricing',
        auth: { kind: 'token' },
        windows: [{ id: 'hypercredits', label: 'Credits', color: CREDITS_COLOR }],
        tiers: ['free', 'shred'],
      };
    },

    configure(cfg = {}) {
      if (cfg.api_key !== undefined) apiKey = cfg.api_key ? String(cfg.api_key).trim() : null;
      if (cfg.api_token !== undefined) apiKey = cfg.api_token ? String(cfg.api_token).trim() : null;
      if (cfg.token !== undefined) apiKey = cfg.token ? String(cfg.token).trim() : apiKey;
      // Cookie for the billing page (refresh date scrape). Optional.
      if (cfg.cookie !== undefined) cookie = cfg.cookie;
    },

    async setAuth(payload) {
      apiKey = String(payload ?? '').trim() || null;
    },

    async fetch() {
      if (!apiKey) throw new AuthExpiredError('no hyper API key configured');

      const res = await fetch(API_URL, {
        headers: {
          Authorization: `Bearer ${apiKey}`,
          'User-Agent': USER_AGENT,
          Accept: 'application/json',
        },
      });

      if (res.status === 401 || res.status === 403) throw new AuthExpiredError();
      if (res.status === 429) {
        throw new RateLimitedError(Number(res.headers.get('retry-after')) || null);
      }
      if (!res.ok) throw new Error(`hyper.charm.land HTTP ${res.status}`);

      const text = await res.text();
      const parsed = parse(text);

      // Best-effort: fetch the billing page to get the refresh date.
      if (cookie) {
        try {
          const billRes = await fetch(BILLING_URL, {
            headers: {
              Cookie: cookie,
              'User-Agent': BROWSER_UA,
              Accept: 'text/html',
            },
            redirect: 'manual',
            signal: AbortSignal.timeout(5000),
          });
          if (billRes.ok) {
            const billHtml = await billRes.text();
            const resetsAt = parseRefreshText(billHtml);
            if (resetsAt && parsed.windows?.[0]) {
              parsed.windows[0].resets_at = resetsAt;
            }
          }
        } catch {
          // billing page fetch is best-effort — resets_at stays null
        }
      }

      return text;
    },

    intervalSeconds() {
      return 300;
    },

    parse,
  };
}

export { createProvider };

File: usage-daemon-v3/usage_daemon/providers/hyper.py
"""Charm Hyper credits provider plugin (port of src/providers/hyper.js,
extended per the locked data model: tier + per-model usage from the dashboard).

Two sources:
  1. GET /v1/credits (Bearer API key)  -> {"balance": 78}
  2. GET /teams/<id>/dashboard (cookie) -> "Next Hypercredit Refresh in 2 weeks"
     (resets_at) + Plan card (tier) + "Usage by Model" table (segments).
     Optional; absent cookie => resets_at null, tier free, no segments.

Free tier: 200 credits/month (hardcoded — API doesn't return cap; verified on
the dashboard 2026-09). Used = cap - balance.
"""

from __future__ import annotations

import json
import re
import time as _time
from datetime import datetime, timezone

from ..errors import AuthExpiredError, RateLimitedError
from ..httputil import create_client

CREDITS_COLOR = "#0072B2"

ID = "hyper"
LABEL = "Charm Hyper"

CREDITS_URL = "https://hyper.charm.land/v1/credits"
DASHBOARD_URL = (
    "https://hyper.charm.land/teams/6b4e808f-3933-4bfd-a321-be1bcfe04f21/dashboard"
)
USER_AGENT = "usage-daemon"
BROWSER_UA = "Mozilla/5.0 (X11; Linux x86_64; rv:153.0) Gecko/20100101 Firefox/153.0"

PLAN_CAP = 200  # free tier: 200 credits/month


def parse_refresh_text(html: str) -> str | None:
    """Parse 'Next Hypercredit Refresh in 4 weeks' -> ISO timestamp."""
    m = re.search(
        r"Next Hypercredit Refresh in (\d+)\s+(day|days|week|weeks)", html, re.I
    )
    if not m:
        return None
    n = int(m.group(1))
    unit = m.group(2).lower()
    days = n * 7 if unit.startswith("week") else n
    ts = _time.time() + days * 24 * 3600
    return datetime.fromtimestamp(ts, timezone.utc).isoformat()


def parse_dashboard(html: str) -> dict:
    """Extract {tier, resets_at, segments} from the dashboard page (best-effort)."""
    res: dict = {}
    m = re.search(r">\s*Plan\s*</p>\s*<p[^>]*>\s*([A-Za-z]+)\s*<", html, re.S)
    if m:
        res["tier"] = m.group(1).lower()
    reset = parse_refresh_text(html)
    if reset:
        res["resets_at"] = reset

    segments = []
    for block in re.findall(r'<tr class="hover[^"]*">(.*?)</tr>', html, re.S):
        cells = re.findall(r"<td[^>]*>(.*?)</td>", block, re.S)
        if len(cells) < 7:
            continue
        mname = re.search(r"<span[^>]*>([^<]+)</span>", cells[0])
        if not mname:
            continue

        def _num(td: str):
            mm = re.sub(r"[^0-9.]", "", td)
            return float(mm) if mm else None

        segments.append({
            "model": mname.group(1).strip(),
            "input_tokens": _num(cells[1]),
            "output_tokens": _num(cells[2]),
            "cache_tokens": _num(cells[3]),
            "total_tokens": _num(cells[4]),
            "hypercredits": _num(cells[5]),
            "cost_usd": _num(cells[6]),
        })
    if segments:
        res["segments"] = segments
    return res


def parse(raw) -> dict:
    """Pure function of the credits JSON text -> {tier, windows, segments, _hyper}."""
    try:
        data = json.loads(raw) if isinstance(raw, str) else raw
    except Exception:
        raise AuthExpiredError("unparseable hyper response")
    if not data or not isinstance(data, dict):
        raise AuthExpiredError("unparseable hyper response")

    if data.get("error"):
        err = data["error"]
        if err.get("type") == "authentication_error":
            raise AuthExpiredError(err.get("message") or "invalid hyper API key")
        raise RuntimeError(f"hyper API error: {err.get('message') or json.dumps(err)}")

    balance = data.get("balance")
    if not isinstance(balance, (int, float)) or isinstance(balance, bool):
        raise AuthExpiredError("no usable hypercredits balance")

    used = max(0, PLAN_CAP - balance)
    pct = max(0, min(100, (100 * used) / PLAN_CAP))

    windows = [
        {
            "id": "hypercredits",
            "label": "Credits",
            "letter": "Hc",
            "pct": pct,
            "used": used,
            "cap": PLAN_CAP,
            "unit": "credits",
            "resets_at": None,  # filled in from the dashboard, best-effort
            "color": CREDITS_COLOR,
            "will_deplete": False,
        }
    ]
    return {
        "tier": "free",
        "windows": windows,
        "segments": [],
        "_hyper": {"balance": balance},
    }


def create_provider(client=None):
    st = {
        "apiKey": None,
        "cookie": None,
        "client": client,
        "dashboard_url": DASHBOARD_URL,
    }

    def config():
        return {
            "id": ID,
            "label": LABEL,
            "usageUrl": DASHBOARD_URL,
            "auth": {"kind": "token"},
            "windows": [
                {"id": "hypercredits", "label": "Credits", "color": CREDITS_COLOR}
            ],
            "tiers": ["free", "shred"],
        }

    def configure(cfg: dict | None = None):
        cfg = cfg or {}
        if "api_key" in cfg:
            st["apiKey"] = cfg["api_key"].strip() if cfg["api_key"] else None
        if "api_token" in cfg:
            st["apiKey"] = cfg["api_token"].strip() if cfg["api_token"] else None
        if "token" in cfg:
            st["apiKey"] = cfg["token"].strip() if cfg["token"] else st["apiKey"]
        if "cookie" in cfg:
            st["cookie"] = cfg["cookie"]

    async def set_auth(payload: str) -> None:
        st["apiKey"] = (payload or "").strip() or None

    async def fetch() -> str:
        if not st["apiKey"]:
            raise AuthExpiredError("no hyper API key configured")
        c = st["client"] or create_client()
        res = await c.get(
            CREDITS_URL,
            headers={
                "Authorization": f"Bearer {st['apiKey']}",
                "User-Agent": USER_AGENT,
                "Accept": "application/json",
            },
        )
        if res.status_code in (401, 403):
            raise AuthExpiredError()
        if res.status_code == 429:
            ra = res.headers.get("retry-after")
            raise RateLimitedError(int(ra) if ra and ra.isdigit() else None)
        if res.status_code >= 400:
            raise RuntimeError(f"hyper.charm.land HTTP {res.status_code}")
        return res.text

    async def fetch_dashboard() -> str:
        """Best-effort dashboard fetch (cookie); '' if unreachable/unauthed."""
        if not st["cookie"]:
            return ""
        c = st["client"] or create_client()
        try:
            res = await c.get(
                st["dashboard_url"],
                headers={
                    "Cookie": st["cookie"],
                    "User-Agent": BROWSER_UA,
                    "Accept": "text/html",
                },
            )
            return res.text if res.status_code < 400 else ""
        except Exception:
            return ""

    def interval_seconds():
        return 300

    return {
        "id": ID,
        "label": LABEL,
        "auth": {"kind": "token"},
        "config": config,
        "configure": configure,
        "set_auth": set_auth,
        "fetch": fetch,
        "fetch_dashboard": fetch_dashboard,
        "parse_dashboard": parse_dashboard,
        "interval_seconds": interval_seconds,
        "parse": parse,
    }